### PR TITLE
add MegaChemicalReactor and MegaOilCracker

### PR DESCRIPTION
**Pull request by SKYCATV587**

### CODE REVIEW

**Review by draknyte1 — COMMENTED** [review]
I think we should either nerf the LCR if this is added with the ability to be properly tiered or just discourage further use in general.

**Comment by Steelux8** [comment]
From the screenshot, it looks like a pretty small multi to be a mega. On the other hand, avoiding those abominations that never fit in any build is also something I'd like.

**Comment by bombcar** [comment]
Perhaps rename it the Very Large Chemical Reactor and make it a bit bigger heh, instead of a 3x3 glass face 5x5?

**Comment by SKYCATV587** [comment]
> Perhaps rename it the Very Large Chemical Reactor and make it a bit bigger heh, instead of a 3x3 glass face 5x5?

Because mcr is different from other mega machines, it will be used more, so I don't design it very large

File: src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/ItemRegistry.java [hunk 1]
```diff
@@ -215,6 +217,9 @@ public static void run() {
             ItemRegistry.compressedHatch = new GT_MetaTileEntity_CompressedFluidHatch(ConfigHandler.IDOffset + GT_Values.VN.length * 8 + 8, "CompressedFluidHatch", "Liquid Air Fluid Hatch").getStackForm(1L);
             ItemRegistry.giantOutputHatch = new GT_MetaTileEntity_GiantOutputHatch(ConfigHandler.IDOffset + GT_Values.VN.length * 8 + 9, "GiantOutputHatch", "Giant Output Hatch").getStackForm(1L);
             ItemRegistry.megaMachines[2] = new GT_TileEntity_MegaDistillTower(ConfigHandler.IDOffset + GT_Values.VN.length * 8 + 10, "MegaDistillationTower", "Mega Distillation Tower").getStackForm(1L);
+            ItemRegistry.megaMachines[3] = new GT_TileEntity_MegaChemicalReactor(13111, "MegaChemicalReactor", "Mega Chemical Reactor").getStackForm(1L);
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/loaders/ItemRegistry.java]: Where did you get these IDs from? Put it in bartworks' space.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    }
+
+    public GT_TileEntity_MegaChemicalReactor(String aName) {
+        super(aName);
+    }
+
+    @Override
+    public GT_Multiblock_Tooltip_Builder createTooltip() {
+        final GT_Multiblock_Tooltip_Builder tt = new GT_Multiblock_Tooltip_Builder();
+        tt.addMachineType("Molecular Surgery Machine")
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java]: Don't get too creative at machine type. It makes searching in NEI harder than it should be. Let's keep it the same as the non mega version.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+                .addInfo("Controller block for the Chemical Reactor")
+                .addInfo("What molecule do you want to synthesize")
+                .addInfo("Or you want to replace something in this molecule")
+                .addInfo("The structure is too complex!")
+                .addInfo("Follow the Structure Lib hologram projector to build the main structure.")
+                .addSeparator()
+                .beginStructureBlock(5, 5,  9,  false)
+                .addController("Front bottom")
+                .addEnergyHatch("Hint block with dot 3")
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java]: Use `.addXXX("Location of hatch, Descriptive or not", 3)` instead.

For energy hatches specifically, should also notice the energy hatch tier cannot exceed glass tier.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+                .addMaintenanceHatch("Hint block with dot 2")
+                .addInputHatch("Hint block with dot 1")
+                .addInputBus("Hint block with dot 1")
+                .addOutputBus("Hint block with dot 1")
+                .addOutputHatch("Hint block with dot 1")
+                .addStructureHint("This Mega Multiblock is too big to have its structure hologram displayed fully.")
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java]: Remove this line. 225 blocks is well below the limit.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    public void construct(ItemStack aStack, boolean aHintsOnly) {
+        buildPiece(STRUCTURE_PIECE_MAIN,aStack,aHintsOnly,2,2,0);
+    }
+
+    // -------------- TEC TECH COMPAT ----------------
+
+    @Override
+    public boolean checkMachine(IGregTechTileEntity aBaseMetaTileEntity, ItemStack aStack) {
+        glasTier = 0;
+        if (this.glasTier != 8 && !this.mEnergyHatches.isEmpty()) {
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java:210]: You will probably want this if block to happen after checkPiece().

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+            .addElement('g', ofChain(
+                    ofBlockAdder(GT_TileEntity_MegaChemicalReactor::addGlas, ItemRegistry.bw_glasses[0], 1)
+            ))
+            .build();
+
+    private int mCasingAmount;
+    private int mCoilAmount;
+
+    @Override
+    public String[] getStructureDescription(ItemStack itemStack){
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java]: If you used addXXX in createTooltip properly you don't need to override this.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+            .addElement('c', ofChain(
+                    onElementPass(GT_TileEntity_MegaChemicalReactor::onCoilAdded, ofBlock(GregTech_API.sBlockCasings4, 7))
+            ))
+            .addElement('g', ofChain(
+                    ofBlockAdder(GT_TileEntity_MegaChemicalReactor::addGlas, ItemRegistry.bw_glasses[0], 1)
+            ))
+            .build();
+
+    private int mCasingAmount;
+    private int mCoilAmount;
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java]: mCoilAmount is completely unused. Remove it.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+            ))
+            .addElement('r', ofChain(
+                    ofHatchAdder(GT_TileEntity_MegaChemicalReactor::addMaintenanceToMachineList, CASING_INDEX, 2)
+            ))
+            .addElement('e', ofChain(
+                    ofHatchAdder(GT_TileEntity_MegaChemicalReactor::addEnergyInputToMachineList, CASING_INDEX, 3),
+                    onElementPass(GT_TileEntity_MegaChemicalReactor::onCasingAdded, ofBlock(GregTech_API.sBlockCasings8, 0))
+            ))
+            .addElement('c', ofChain(
+                    onElementPass(GT_TileEntity_MegaChemicalReactor::onCoilAdded, ofBlock(GregTech_API.sBlockCasings4, 7))
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java]: mCoilAmount is completely unused. Remove it.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+                "Hint block with dot 3：EnergyHatch"
+        };
+    }
+
+    @Override
+    public IStructureDefinition<GT_MetaTileEntity_LargeChemicalReactor> getStructureDefinition() {
+        return (IStructureDefinition) STRUCTURE_DEFINITION;
+    }
+
+    private void onCasingAdded() {
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaChemicalReactor.java:263]: mCoilAmount is completely unused. Remove it.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaOilCracker.java [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+                .addInfo("Gives different benefits whether it hydro or steam-cracks:")
+                .addInfo("Hydro - Consumes 20% less Hydrogen and outputs 25% more cracked fluid")
+                .addInfo("Steam - Outputs 50% more cracked fluid")
+                .addInfo("(Values compared to cracking in the Chemical Reactor)")
+                .addInfo("Place the appropriate circuit in the controller")
+                .addSeparator()
+                .beginStructureBlock(13, 7, 9, true)
+                .addController("Front bottom")
+                .addInfo("Gets 10% EU/t reduction per coil tier, up to a maximum of 50%")
+                .addEnergyHatch("Hint block with dot 3")
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaOilCracker.java]: Same as the other mult here.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaOilCracker.java [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+    public void construct(ItemStack aStack, boolean aHintsOnly) {
+        buildPiece(STRUCTURE_PIECE_MAIN,aStack,aHintsOnly,6,6,0);
+    }
+
+    // -------------- TEC TECH COMPAT ----------------
+
+    @Override
+    public boolean checkMachine(IGregTechTileEntity aBaseMetaTileEntity, ItemStack aStack) {
+        glasTier = 0;
+        if (this.glasTier != 8 && !this.mEnergyHatches.isEmpty()) {
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaOilCracker.java:246]: Same here.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaOilCracker.java [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+import com.github.bartimaeusnek.bartworks.util.BW_Util;
+import com.github.bartimaeusnek.bartworks.util.MegaUtils;
+import com.github.bartimaeusnek.bartworks.util.Pair;
+import com.github.bartimaeusnek.crossmod.tectech.TecTechEnabledMulti;
+import com.github.bartimaeusnek.crossmod.tectech.helper.TecTechUtils;
+import com.gtnewhorizon.structurelib.StructureLibAPI;
+import com.gtnewhorizon.structurelib.structure.IStructureDefinition;
+import com.gtnewhorizon.structurelib.structure.IStructureElement;
+import com.gtnewhorizon.structurelib.structure.StructureDefinition;
+//import com.sun.istack.internal.NotNull;
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaOilCracker.java]: Remove commented out imports.

File: src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaOilCracker.java [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+        return new String[]{
+                "Hint block with dot 1: EnergyHatch,MaintenanceHatch",
+                "Hint block with dot 2: InputHatch/OutputHatch",
+                "Hint block with dot 3: OutputHatch/InputHatch",
+                "Hint block with dot 4: Steam/Hydrogen InputHatch"
+        };
+    }
+
+    @Override
+    public HeatingCoilLevel getCoilLevel() {
```
Comment (Glease) [inline src/main/java/com/github/bartimaeusnek/bartworks/common/tileentities/multis/mega/GT_TileEntity_MegaOilCracker.java]: You don't need to override this. Just use `this.getCoilLever()` wherever super class uses `this.heatLevel`

**Review by Glease — CHANGES_REQUESTED** [review]
Balance aside, there is a lot of minor issues with code. Much of the review comments in mega LCR also applies to the mega cracker, but I got lazy..😛

**Comment by Dream-Master** [comment]
@SKYCATV587

**Comment by Steelux8** [comment]
Is the mega LCR size fine? What's the size of it now? It seems quite smaller than the other megas, including this new oil cracker. With how overpowered the LCR is, I'm not sure about this.

**Comment by SKYCATV587** [comment]
> Is the mega LCR size fine? What's the size of it now? It seems quite smaller than the other megas, including this new oil cracker. With how overpowered the LCR is, I'm not sure about this.

Yes ,Because mcr is different from other mega machines, it will be used more, so I don't design it very large,I only let it be more expensive than other MEGA Machine,it need 7 fusion coils.

**Review by Glease — APPROVED** [review]
Code wise it now looks ok. Not sure about balance though.

**Comment by Steelux8** [comment]
Balance can be changed later, and it heavily depends on whether or not the LCR will get rebalanced as well.

**Comment by Dream-Master** [comment]
> Balance can be changed later, and it heavily depends on whether or not the LCR will get rebalanced as well.

No balance need to be done first.
Sure we can do a zeta test first and open up discussion later

**Comment by Pxx500** [comment]
image
Quality design, do you even test before implementing?

**Comment by Pxx500** [comment]
image
Same here, good design

**Comment by SKYCATV587** [comment]
> image Same here, good design

MBF and MDT have the same bug,i dont know how to deal with it

**Comment by kuba6000** [comment]
> > image Same here, good design
> 
> MBF and MDT have the same bug,i dont know how to deal with it

I fixed it already here: https://github.com/GTNewHorizons/bartworks/pull/97
The MBF and MDT works normally.

**Comment by SKYCATV587** [comment]
> > > image Same here, good design
> > 
> > 
> > MBF and MDT have the same bug,i dont know how to deal with it
> 
> I fixed it already here: #97 The MBF and MDT works normally.

oh, thanks， i saw it .I just thought it was another bug before I said mbf and mdt had

**Pull request merged**